File: system-test/typescript/src/testkit/driver/adapter/external/erp/erp-real-driver.ts
import type { Result } from '../../../../common/result.js';
import { success } from '../../../../common/result.js';
import type { ErpErrorResponse } from '../../../port/external/erp/dtos/ErpErrorResponse.js';
import type { GetProductResponse } from '../../../port/external/erp/dtos/GetProductResponse.js';
import type { ReturnsProductRequest } from '../../../port/external/erp/dtos/ReturnsProductRequest.js';
import type { ReturnsPromotionRequest } from '../../../port/external/erp/dtos/ReturnsPromotionRequest.js';
import type { ErpDriver } from '../../../port/external/erp/erp-driver.js';
import { ErpRealClient } from './client/ErpRealClient.js';

export class ErpRealDriver implements ErpDriver {
  private readonly client: ErpRealClient;

  constructor(baseUrl: string) {
    this.client = new ErpRealClient(baseUrl);
  }

  async goToErp(): Promise<Result<void, ErpErrorResponse>> {
    return this.client.checkHealth();
  }

  async getProduct(sku: string): Promise<Result<GetProductResponse, ErpErrorResponse>> {
    return this.client.getProduct(sku);
  }

  async returnsProduct(request: ReturnsProductRequest): Promise<Result<void, ErpErrorResponse>> {
    return this.client.createProduct(request);
  }

  async returnsPromotion(_request: ReturnsPromotionRequest): Promise<Result<void, ErpErrorResponse>> {
    return success(undefined);
  }

  async close(): Promise<void> {}
}

File: system/multitier/backend-java/src/testSupport/java/com/mycompany/myshop/backend/testkit/driver/adapter/external/erp/ErpRealDriver.java
package com.mycompany.myshop.backend.testkit.driver.adapter.external.erp;

import com.mycompany.myshop.backend.testkit.driver.adapter.external.erp.client.ErpRealClient;
import com.mycompany.myshop.backend.testkit.driver.port.external.erp.ErpDriver;

/**
 * ERP driver backed by the real simulator via {@link ErpRealClient}, mirroring system-test's own
 * {@code ErpRealDriver}. {@code returnsProduct} provisions the fixture for real, since there is nothing
 * to stub. {@code returnsNoProduct}/{@code returnsPromotion} are no-ops: a SKU that was never created is
 * already absent, and the simulator's promotion endpoint is hardcoded, so there is nothing to arrange —
 * same shape as system-test's {@code ErpRealDriver.returnsPromotion}. {@code failsForProduct}/
 * {@code failsForPromotion} have no real counterpart at all — the simulator cannot be told to fail on
 * demand — so they throw rather than silently succeed, which would make an error-path test pass for the
 * wrong reason.
 */
public class ErpRealDriver implements ErpDriver {

    private final ErpRealClient client;

    public ErpRealDriver(String baseUrl) {
        this.client = new ErpRealClient(baseUrl);
    }

    @Override
    public void goToErp() {
        client.checkHealth();
    }

    @Override
    public void returnsProduct(String sku, String price) {
        client.createProduct(sku, price);
    }

    @Override
    public void returnsNoProduct(String sku) {
        // no-op: a SKU that was never created is already absent
    }

    @Override
    public void returnsPromotion(boolean active, String discount) {
        // no-op: the simulator's promotion endpoint is hardcoded, nothing to arrange
    }

    @Override
    public void failsForProduct(String sku, int status, String body) {
        throw new UnsupportedOperationException(
            "ErpRealDriver cannot force a failure response — the real ERP simulator has no fault-injection endpoint.");
    }

    @Override
    public void failsForPromotion(int status, String body) {
        throw new UnsupportedOperationException(
            "ErpRealDriver cannot force a failure response — the real ERP simulator has no fault-injection endpoint.");
    }
}
